### PR TITLE
ENH: EA._hasnans, EADtype.empty

### LINKED ISSUE #19600: Add ExtesnsionDtype.empty classmethod
This would let us provide a default `construct_from_string` method.

> It seems we could have a default implementation for ExtensionDtype.construct_from_string ? (I now just copy pasted from the decimal example, and I think json example also has the same basic one)

The default would have to rely on `ExtensionDtype()` being constructable with no arguments, or an `ExtensionDtype.empty` method. For many that'll be

```python
@classmethod
def empty(cls):
    return cls()
```

which is easier than implementing `construct_from_string`. Maybe call it `from_empty`.

### LINKED ISSUE #22680: Add a method to ExtensionArray interface for whether an array contains NAs
In lots of places, pandas does something like `if np.any(arr.isna())`, which is wasteful as we have to create an ndarray of booleans just to check whether there are any `True` values.

Some arrays, like Arrow, know ahead of time whether there are any NAs in the array. Would it make sense to expose an API for an array saying whether they have any missing values?

With indexes, we work around this by caching a `_hasnans` value. That wouldn't work for mutable arrays.

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #19600
- [x] closes #22680
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/base.py [hunk 1]
```diff
@@ -603,6 +603,16 @@ def isna(self) -> np.ndarray | ExtensionArraySupportsAnyAll:
         """
         raise AbstractMethodError(self)
 
+    @property
```
Comment (jreback) [inline pandas/core/arrays/base.py:606]: can we not use a cached_property?

**Review by jreback — CHANGES_REQUESTED** [review]
small comment

File: pandas/core/arrays/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/base.py:606]: no, EAs are mutable

Comment (jreback) [reply pandas/core/arrays/base.py:606]: right (though actually we *could* do this as we know when they *are* mutated, e.g keep a dirty bit). but out of scope

Comment (jbrockmendel) [reply pandas/core/arrays/base.py:606]: i've looked into this w/r/t `DatetimeArray.freq` and found that propagating it through views (e.g. `arr[:5][0] = np.nan`) makes this a giant pain

**Review by jreback — APPROVED** [review]

File: pandas/tests/extension/base/constructors.py [hunk 2]
```diff
@@ -126,6 +126,12 @@ def test_construct_empty_dataframe(self, dtype):
...
-
         assert isinstance(result, cls)
         assert result.dtype == dtype
+        assert result.shape == (4,)
+
+        # GH#19600 method on ExtensionDtype
+        result2 = dtype.empty((4,))
+        assert isinstance(result2, cls)
+        assert result2.dtype == dtype
+        assert result2.shape == (4,)
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/base/constructors.py:137]: If we make this public, we should maybe a test that passing a scalar integer works as well (if we want to allow that, following np.empty)

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for adding this! A couple of comments though.

Can you add a reference to this public EADtype.empty from ExtensionArray._empty (which you added in https://github.com/pandas-dev/pandas/pull/40602), indicating that EADtype.empty is the official public variant people should use?

File: pandas/core/arrays/base.py [hunk 3]
```diff
@@ -603,6 +603,16 @@ def isna(self) -> np.ndarray | ExtensionArraySupportsAnyAll:
...
+    def _hasnans(self) -> bool:
+        # GH#22680
+        """
+        Equivalent to `self.isna().any()`.
+
+        Some ExtensionArray subclasses may be able to optimize this check.
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/base.py:612]: > I would also rename it to _hasna or so, to be consistent with our public NA related properties 

I used _hasnans to be consistent with the datetimelike arrays (didn't realize the masked arrays had _hasna until just now).  Also consistent-ish with the hasnans we have in Index and Series.

We should revisit the naming convention(s) once the NA/NaN API design discussion reaches some kind of conclusion.

> (we could also make this public?)

I chose to keep this private so we would have the option of changing it later.

(Side-note: "private" for EA methods is really ambiguous, could mean a) user shouldn't use use, b) should only be called from `self`, c) 3rd party authors shouldn't use, d) 3rd party implementors shouldn't override.  Not sure how we can clarify or tighten this)

File: pandas/core/dtypes/base.py [hunk 4]
```diff
@@ -208,6 +209,19 @@ def construct_array_type(cls) -> type_t[ExtensionArray]:
         """
         raise AbstractMethodError(cls)
 
+    def empty(self, shape: Shape) -> type_t[ExtensionArray]:
+        """
+        Construct an ExtensionArray of this dtype with the given shape.
+
+        Analogous to numpy.empty.
+
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/dtypes/base.py:217]: sure, just added this to me next follow-ups branch.

File: pandas/core/arrays/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/core/arrays/base.py:612]: The current situation is IMO implicitly: everything in the base class is automatically "public" for developers of third party EAs, unless explicitly documented.

`_hasnans` is being used in methods that EA authors can override, and is useful for them as well, and they are also allowed to override it. So I think this is "public", and thus should decide on a name now, IMO.

**Pull request merged**